### PR TITLE
Add complex support to parametrizations.spectral_norm

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121452

Fixes https://github.com/pytorch/pytorch/issues/121091

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @amjames

### CODE REVIEW

File: test/nn/test_parametrization.py [hunk 1]
```diff
@@ -1392,16 +1392,17 @@ def test_new_spectral_norm_forward(self):
...
-            # weights should be rescaled by spectral norm, so should now be
-            # diag(1., 0.5)
-            expected = torch.diag(torch.tensor([1.0, 0.5]))
-            self.assertEqual(m.weight.data, expected)
+        for dtype in (torch.float, torch.cfloat):
+            m = nn.Linear(2, 2, dtype=dtype)
+            with torch.no_grad():
+                # set weight to diagonal matrix with values 2., 1.
+                # so spectral norm is 2
+                m.weight = nn.Parameter(torch.diag(torch.tensor([2.0, 1.0], dtype=dtype)))
```
Comment (peterbell10) [inline test/nn/test_parametrization.py:1400]: Can you test with a complex-valued weight.

**Review by ezyang — APPROVED** [review]

File: test/nn/test_parametrization.py (hunk 1 above)
Comment (lezcano) [reply test/nn/test_parametrization.py:1400]: this is a complex-valued weight?

Comment (lezcano) [reply test/nn/test_parametrization.py:1400]: ah I see what you mean. Sure.

**Review by peterbell10 — APPROVED** [review]

**Pull request closed without merging**